consensus/istanbul/proxy/val_enodes_share.go

```diff
@@ -0,0 +1,158 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the celo library. If not, see <http://www.gnu.org/licenses/>.
+
+package proxy
+
+import (
+	"encoding/hex"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+
+func (p *proxyEngine) generateValEnodesShareMsg(remoteValidators []common.Address) (*istanbul.Message, error) {
+	vetEntries, err := p.backend.GetValEnodeTableEntries(remoteValidators)
+
+	if err != nil {
+		p.logger.Error("Error in retrieving all the entries from the ValEnodeTable", "err", err)
+		return nil, err
+	}
+
+	sharedValidatorEnodes := make([]sharedValidatorEnode, 0, len(vetEntries))
+	for address, vetEntry := range vetEntries {
+		if vetEntry.GetNode() == nil {
+			continue
+		}
+		sharedValidatorEnodes = append(sharedValidatorEnodes, sharedValidatorEnode{
+			Address:  address,
+			EnodeURL: vetEntry.GetNode().String(),
+			Version:  vetEntry.GetVersion(),
+		})
+	}
+
+	valEnodesShareData := &valEnodesShareData{
+		ValEnodes: sharedValidatorEnodes,
+	}
+
+	valEnodesShareBytes, err := rlp.EncodeToBytes(valEnodesShareData)
+	if err != nil {
+		p.logger.Error("Error encoding Istanbul Validator Enodes Share message content", "ValEnodesShareData", valEnodesShareData.String(), "err", err)
+		return nil, err
+	}
+
+	msg := &istanbul.Message{
+		Code:      istanbul.ValEnodesShareMsg,
+		Msg:       valEnodesShareBytes,
+		Address:   p.address,
+		Signature: []byte{},
+	}
+
+	p.logger.Trace("Generated a Istanbul Validator Enodes Share message", "IstanbulMsg", msg.String(), "ValEnodesShareData", valEnodesShareData.String())
+
+	return msg, nil
+}
+
+func (p *proxyEngine) SendValEnodesShareMsg(proxyPeer consensus.Peer, remoteValidators []common.Address) error {
+	logger := p.logger.New("func", "sendValEnodesShareMsg")
+
+	msg, err := p.generateValEnodesShareMsg(remoteValidators)
+	if err != nil {
+		logger.Error("Error generating Istanbul ValEnodesShare Message", "err", err)
+		return err
+	}
+
+	// Sign the validator enode share message
+	if err := msg.Sign(p.backend.Sign); err != nil {
+		logger.Error("Error in signing an Istanbul ValEnodesShare Message", "ValEnodesShareMsg", msg.String(), "err", err)
+		return err
+	}
+
+	// Convert to payload
+	payload, err := msg.Payload()
+	if err != nil {
+		logger.Error("Error in converting Istanbul ValEnodesShare Message to payload", "ValEnodesShareMsg", msg.String(), "err", err)
+		return err
+	}
+
+	logger.Trace("Sending Istanbul Validator Enodes Share payload to proxy peer", "proxyPeer", proxyPeer)
+	if err := proxyPeer.Send(istanbul.ValEnodesShareMsg, payload); err != nil {
```

Comment:
i think a nicer abstractin would be to have an `IstanbulPeer` that has a `func SendValEnodesShares()` the same way there's an EthereumPeer with similar methods